### ISSUE TITLE
h3 do not send CONNECTION_CLOSE frame on lossy connections

### ISSUE THREAD

**Issue by IcebrakerCZ** [litespeedtech/lsquic#496]
Hi guys, we are using the http_client in Akamai to check compatibility with our quic server implementation.

Lately we can see a problem in the communication that the http_client do not close connection after everything is successfully processed on connections with 2% loss on 256kbit/s bandwidth when 1MB file is downloaded in single connection with single h3 request (tried on lsquic master branch too).

I found in debug log that CONNECTION_CLOSE frame is tried to be sent but is not sent due to cwnd and/or pacer and the connection is destroyed without waiting on the CONNECTION_CLOSE frame being actually sent.

I tried to debug it and it looks like to me that when bbr starts to be used as congestion control and two packets (72 bytes in total only) are already lost cwnd drops from 48KB to 4*TCPMSS (5840 bytes only) and CONNECTION_CLOSE can't be sent because cwnd is full:

http_client starts at 15:53:19.505

1) bandwidth drop

15:53:26.410
bbr: becoming application-limited.  Last sent packet: 220; CWND: 48124
sendctl: send_ctl_can_send: n_out: 7588 (unacked_all: 7588); cwnd: 48124; ccfc: 56/1048576
sendctl: send_ctl_can_send: n_out: 7588 (unacked_all: 7588); cwnd: 48124; ccfc: 56/1048576
bbr: becoming application-limited.  Last sent packet: 221; CWND: 5840
bbr: becoming application-limited.  Last sent packet: 222; CWND: 5840
sendctl: send_ctl_can_send: n_out: 5443 (unacked_all: 5443); cwnd: 5840; ccfc: 56/1048576
sendctl: send_ctl_can_send: n_out: 5443 (unacked_all: 5443); cwnd: 5840; ccfc: 56/1048576
bbr: becoming application-limited.  Last sent packet: 223; CWND: 5840

2) CONECTION_CLOSE can't be sent due to full cwnd

15:53:54.474
pacer: lsquic_pacer_can_schedule: 1
sendctl: pacer blocked: 0, in_flight_all: 778
sendctl: send_ctl_can_send: n_out: 36513 (unacked_all: 36513); cwnd: 5840; ccfc: 56/1048576
sendctl: send_ctl_can_send: can't send - pacing - cwnd full    <- log message aded by me
sendctl: send_ctl_can_send: n_out: 36513 (unacked_all: 36513); cwnd: 5840; ccfc: 56/1048576
sendctl: send_ctl_can_send: can't send - pacing - cwnd full
conn: connection is OK to close
sendctl: send_ctl_can_send: n_out: 36513 (unacked_all: 36513); cwnd: 5840; ccfc: 56/1048576
sendctl: send_ctl_can_send: can't send - pacing - cwnd full
conn: used up packet allowance, quiet now (line 8559)

3) after 15 seconds a PING and another CONNECTION_CLOSE is tried to be sent

15:54:09.456
alarmset: ring expired PING alarm
conn: Ping alarm rang: schedule PING frame to be generated
pacer: lsquic_pacer_can_schedule: 1
sendctl: pacer blocked: 0, in_flight_all: 779
sendctl: send_ctl_can_send: n_out: 36574 (unacked_all: 36574); cwnd: 5840; ccfc: 56/1048576
sendctl: send_ctl_can_send: can't send - pacing - cwnd full
conn: cannot get writeable packet for PING frame
sendctl: send_ctl_can_send: n_out: 36574 (unacked_all: 36574); cwnd: 5840; ccfc: 56/1048576
sendctl: send_ctl_can_send: can't send - pacing - cwnd full
conn: connection is OK to close
sendctl: send_ctl_can_send: n_out: 36574 (unacked_all: 36574); cwnd: 5840; ccfc: 56/1048576
sendctl: send_ctl_can_send: can't send - pacing - cwnd full
conn: used up packet allowance, quiet now (line 8559)
engine: next advisory tick is 15017717 usec away: conn 4749941DFF5277FA: IDLE

4) after another 15 seconds is the connection destroyed without sending the CONNECTION_CLOSE

15:54:24.489
sendctl: stats: n_total_sent: 921; n_resent: 0; n_delayed: 0
Connection closed.  Status: 4.  Message: <not set>
handshake: iquic_esfi_destroy
conn: # ticks: 1805
conn: sent 921 packets
conn: received 918 packets, of which 0 were not decryptable, 0 were dups and 0 were errors; sent 921 packets, avg stream data per outgoing packet is 0 bytes
conn: ACKs: delayed acks on: no; in: 18; processed: 18; merged: 0
event: full connection destroyed

Well everything has been downloaded correctly and the stream is closed correctly etc. so everything is fine but the timeout looks like something that shouldn't occur right?

I have some observations although I'm not sure how much relevant they are:

- CONNECTION_CLOSE frames are not removed from IQUIC_FRAME_RETX_MASK but maybe it's intentional because the connection is closed after it is sent?

- ACK and CONNECTION_CLOSE frames shouldn't be counted toward congestion control limits but still I download only 1MB file and cwnd is full and CONNECTION_CLOSE can't be sent due to full cwnd as well.

Beside that I mean I'm downloading 1MB file so client should send ACK frames mostly (like probably 99% mostly) because there is no need to send anything else normally so there is no reason for cwnd to be full but it is full?

Maybe ACK and CONNECTION_CLOSE frames sent in non-ack-eliciting packets should be able to be sent even when cwnd is full?

**Comment by litespeedtech** [comment]
Thanks for the detailed analysis. It is likely an issue with lsquic send controller. We will look into this.

**Comment by litespeedtech** [comment]
Can you please try this patch
```
diff --git a/src/liblsquic/lsquic_full_conn_ietf.c b/src/liblsquic/lsquic_full_conn_ietf.c
index fd8501a..2d1de0c 100644
--- a/src/liblsquic/lsquic_full_conn_ietf.c
+++ b/src/liblsquic/lsquic_full_conn_ietf.c
@@ -8574,7 +8574,8 @@ ietf_full_conn_ci_tick (struct lsquic_conn *lconn, lsquic_time_t now)
                     || 0 != lsquic_send_ctl_n_scheduled(&conn->ifc_send_ctl))
                 )
         {
-            RETURN_IF_OUT_OF_PACKETS();
+            /* CONNECTION_CLOSE frame should not be congestion controlled.
+            RETURN_IF_OUT_OF_PACKETS(); */
             generate_connection_close_packet(conn);
             tick |= TICK_SEND|TICK_CLOSE;
         }
```

**Comment by IcebrakerCZ** [comment]
Yep the CONNECTION_CLOSE frame was sent and server closed the connection without timeout :)

**Comment by litespeedtech** [comment]
Should be fixed in 4.0.6 release.
